src/utils/useTests.ts

```diff
@@ -0,0 +1,117 @@
+import { useCallback, useState } from "react";
+
+import {
+  BrowserInfo,
+  ComparisonResult,
+  TestFieldsFragment,
+  TestResult,
+  TestStatus,
+  ViewportInfo,
+} from "../gql/graphql";
+import { aggregateResult } from "./aggregateResult";
+
+type ViewportData = Pick<ViewportInfo, "id" | "name">;
+type BrowserData = Pick<BrowserInfo, "id" | "key" | "name">;
+
+/**
+ * Pick a single test+comparison (by viewport+browser) from a set of tests
+ * for the same story.
+ */
+export function useTests(tests: TestFieldsFragment[]) {
+  const [selectedTest, onSelectTest] = useState<TestFieldsFragment>(tests[0]);
+  const [selectedBrowserId, onSelectBrowserId] = useState<BrowserInfo["id"]>(
+    selectedTest.comparisons[0].browser.id
+  );
+  const selectedComparison = selectedTest.comparisons.find(
+    (comparison) => comparison.browser.id === selectedBrowserId
+  );
+  if (!selectedComparison) {
+    throw new Error(`No comparison found with browser id ${selectedBrowserId}`);
+  }
+
+  const onSelectViewport = useCallback(
+    (viewport: ViewportData) => {
+      const newSelectedTest = tests.find(
+        ({ parameters }) => parameters.viewport.id === viewport.id
+      );
+      if (!newSelectedTest) throw new Error(`No test found with viewport ${viewport.id}`);
+      onSelectTest(newSelectedTest);
+    },
+    [tests]
+  );
+  const onSelectBrowser = useCallback(
+    (browser: BrowserData) => onSelectBrowserId(browser.id),
+    [onSelectBrowserId]
+  );
+
+  const {
+    isInProgress,
+    changeCount,
+    brokenCount,
+    resultsByBrowser,
+    resultsByViewport,
+    viewportInfoById,
+  } = tests.reduce(
+    (acc, test) => {
+      if (test.status === TestStatus.InProgress) {
+        acc.isInProgress = true;
+      }
+      if (test.result === TestResult.Changed) {
+        acc.changeCount += 1;
+      }
+      if (test.result === TestResult.CaptureError || test.result === TestResult.SystemError) {
+        acc.brokenCount += 1;
+      }
+      test.comparisons?.forEach(({ browser, result }) => {
+        acc.resultsByBrowser[browser.id] = aggregateResult([
+          result,
+          acc.resultsByBrowser[browser.id],
+        ]);
+      });
+      test.comparisons?.forEach(({ viewport, result }) => {
+        acc.resultsByViewport[viewport.id] = aggregateResult([
+          result,
+          acc.resultsByViewport[viewport.id],
+        ]);
+      });
+      acc.viewportInfoById[test.parameters.viewport.id] = test.parameters.viewport;
+      return acc;
+    },
+    {
+      isInProgress: false,
+      changeCount: 0,
+      brokenCount: 0,
+      resultsByBrowser: {} as Record<string, ComparisonResult>,
+      resultsByViewport: {} as Record<string, ComparisonResult>,
+      viewportInfoById: {} as Record<string, TestFieldsFragment["parameters"]["viewport"]>,
+    }
+  );
```

Comment:
You can add a generic argument into `reduce` like so:

Warning I wrote this code via GitHub, so formatting is likely a bit off.

```suggestion
  const {
    isInProgress,
    changeCount,
    brokenCount,
    resultsByBrowser,
    resultsByViewport,
    viewportInfoById,
  } = tests.reduce<{
      isInProgress: boolean,
      changeCount: number,
      brokenCount: number,
      resultsByBrowser: Record<string, ComparisonResult>,
      resultsByViewport:Record<string, ComparisonResult>,
      viewportInfoById: Record<string, TestFieldsFragment["parameters"]["viewport"]>,
    }>(
    (acc, test) => {
      if (test.status === TestStatus.InProgress) {
        acc.isInProgress = true;
      }
      if (test.result === TestResult.Changed) {
        acc.changeCount += 1;
      }
      if (test.result === TestResult.CaptureError || test.result === TestResult.SystemError) {
        acc.brokenCount += 1;
      }
      test.comparisons?.forEach(({ browser, result }) => {
        acc.resultsByBrowser[browser.id] = aggregateResult([
          result,
          acc.resultsByBrowser[browser.id],
        ]);
      });
      test.comparisons?.forEach(({ viewport, result }) => {
        acc.resultsByViewport[viewport.id] = aggregateResult([
          result,
          acc.resultsByViewport[viewport.id],
        ]);
      });
      acc.viewportInfoById[test.parameters.viewport.id] = test.parameters.viewport;
      return acc;
    },
    {
      isInProgress: false,
      changeCount: 0,
      brokenCount: 0,
      resultsByBrowser: {},
      resultsByViewport: {},
      viewportInfoById: {},
    }
  );
```

Reply:
Got it. Thanks!